clippy_utils/src/lib.rs

```diff
@@ -2007,6 +2007,10 @@ pub fn is_must_use_func_call(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
 
 /// Checks if an expression represents the identity function
 /// Only examines closures and `std::convert::identity`
+///
+/// Closure bindings with type annotations and `std::convert::identity` with generic args
+/// are not considered identity functions because they can guide type inference,
+/// and removing it may lead to compile errors.
 pub fn is_expr_identity_function(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
```

Comment:
Please note that this utils function might be used from other lints, too (even if it isn't now). So this might inadvertently introduce false negatives. I'd be more comfortable with having the check directly in the lint.

Reply:
This is used only by all the other `*_identity` lints, so would need to be repeated in the other lints as well. So I thought it'd make more sense to have it here

Reply:
A (somewhat arbitrary) example for how the same issue happens in `flat_map_identity`:
```rs
[[]].iter().flat_map(|a: &[i32; 0]| a);
```
clippy recommends replacing it with `.flatten()`, however that fails to compile.

Maybe we could keep the behavior of the current function as it is and have another one with a more obvious name? 🤔

Reply:
I moved this new "breaking" logic into its own util function with a hopefully clear name. The old util function (even if unused now) will continue to behave like it used to, but references the new function in its doc comment because that's likely what one wants to use.
The diff is a bit confusing now because I moved the inner function out so that it can be called from both functions, and the deindenting must be confusing git.

What do you think about this change? There shouldn't and won't be any accidental FNs now this way I think.